### PR TITLE
[ISSUE #900] Support Consul Service Registry

### PR DESCRIPTION
**Pull request by huyuanxin**
Fixes ISSUE #900

### Motivation

make an implement of RegistryService which is support by consul

### Modifications

*Describe the modifications you've done.*

Add ConsulRegistryService and ConsulConstant for implement
Add ConsulRegistryServiceTest for test

### Documentation

- Does this pull request introduce a new feature? (yes / no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
- If a feature is not applicable for documentation, explain why?
- If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by huyuanxin** [comment]
It seems that MacOS 11 build scrip is failed, because could not get resource

**Comment by huyuanxin** [comment]
> seems miss lisence declare

FAILURE: Build failed with an exception.
38 actionable tasks: 5 executed, 33 up-to-date

* What went wrong:
Execution failed for task ':eventmesh-common:compileJava'.
> Could not resolve all files for configuration ':eventmesh-common:compileClasspath'.
   > Could not download log4j-core-2.17.1.jar (org.apache.logging.log4j:log4j-core:2.17.1)
      > Could not get resource 'https://maven.aliyun.com/repository/public/org/apache/logging/log4j/log4j-core/2.17.1/log4j-core-2.17.1.jar'.
         > Premature end of Content-Length delimited message body (expected: 1,790,452; received: 70,916)

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org/

BUILD FAILED in 2m 10s

it seems there is something wrong with gradle build

**Comment by qqeasonchen** [comment]
@xiaoyang-sde please help to review

**Review by xiaoyang-sde — CHANGES_REQUESTED** [review]
Thanks for the contribution! It seems that the changes failed to pass the style check: [https://github.com/apache/incubator-eventmesh/runs/6681779883?check_suite_focus=true#step:5:9859](https://github.com/apache/incubator-eventmesh/runs/6681779883?check_suite_focus=true#step:5:9859). Could you please try to fix this?

**Comment by huyuanxin** [comment]
> 

yep. I will fix it later or tomorrow, work is busy

**Comment by qqeasonchen** [comment]
@huyuanxin any udpates?

**Comment by qqeasonchen** [comment]
@li-xiao-shuang please help to review.

**Comment by huyuanxin** [comment]
already done, but some check still can't be resolved. I can't figure it out

在 2022年6月10日，16:24，Eason Chen ***@***.***> 写道：

﻿

@huyuanxin<https://github.com/huyuanxin> any udpates?

**Comment by huyuanxin** [comment]
Already done. But I can't figure out how to fix the action checks

在 2022年6月10日，16:24，Eason Chen ***@***.***> 写道：

﻿

@huyuanxin<https://github.com/huyuanxin> any udpates?

**Comment by li-xiao-shuang** [comment]
> @li-xiao-shuang please help to review.

Yes, I'm working on it

**Comment by li-xiao-shuang** [comment]
You should create a META-INF/eventmesh/org.apache.eventmesh.api.registry.RegistryService file for the consul plugin so the eventmesh spi will find it

File: eventmesh-registry-plugin/eventmesh-registry-consul/src/main/java/org/apache/eventmesh/registry/consul/service/ConsulRegistryService.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+                    this.consulPort = addr[1];
+                    break;
+                }
+            }
+        }
+    }
+
+    @Override
+    public void start() throws RegistryException {
+        consulClient = new ConsulClient(new ConsulRawClient(consulHost, Integer.parseInt(consulPort), getHttpClient()));
```
Comment (li-xiao-shuang) [inline eventmesh-registry-plugin/eventmesh-registry-consul/src/main/java/org/apache/eventmesh/registry/consul/service/ConsulRegistryService.java:93]: I think ACL should be supported

File: eventmesh-registry-plugin/eventmesh-registry-consul/src/main/java/org/apache/eventmesh/registry/consul/service/ConsulRegistryService.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    @Override
+    public Map<String, Map<String, Integer>> findEventMeshClientDistributionData(String clusterName, String group, String purpose)
+        throws RegistryException {
+        return Collections.emptyMap();
+    }
+
+
+    private HttpClient getHttpClient() {
```
Comment (li-xiao-shuang) [inline eventmesh-registry-plugin/eventmesh-registry-consul/src/main/java/org/apache/eventmesh/registry/consul/service/ConsulRegistryService.java:151]: consul already provides DEFAULT_HTTP_TRANSPORT, I don't think this code is necessary

File: eventmesh-registry-plugin/eventmesh-registry-consul/src/main/java/org/apache/eventmesh/registry/consul/service/ConsulRegistryService.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    @Override
+    public boolean register(EventMeshRegisterInfo eventMeshRegisterInfo) throws RegistryException {
+        try {
+            String[] ipPort = eventMeshRegisterInfo.getEndPoint().split(IP_PORT_SEPARATOR);
+            NewService service = new NewService();
+            service.setPort(Integer.parseInt(ipPort[1]));
+            service.setAddress(ipPort[0]);
+            service.setName(eventMeshRegisterInfo.getEventMeshClusterName());
```
Comment (li-xiao-shuang) [inline eventmesh-registry-plugin/eventmesh-registry-consul/src/main/java/org/apache/eventmesh/registry/consul/service/ConsulRegistryService.java:110]: Many service instances may share the same logical service name. The service name of eventmesh should be used here, not the cluster name. Can the namespace of consul distinguish clusters?

**Comment by li-xiao-shuang** [comment]
How does the heartbeat of the consul service work? Need manual maintenance?

**Comment by qqeasonchen** [comment]
@huyuanxin, please fix the issues that reqeust by @li-xiao-shuang

**Comment by huyuanxin** [comment]
ok， I will fix it later---- Replied Message ----FromEason ***@***.***>Date06/13/2022 19:28 ***@***.***> CcYuanXin ***@***.******@***.***>SubjectRe: [apache/incubator-eventmesh] Support Consul Service Registry (PR #901)
@huyuanxin, please fix the issues that reqeust by @li-xiao-shuang

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/apache/incubator-eventmesh/pull/901#issuecomment-1153800207",
"url": "https://github.com/apache/incubator-eventmesh/pull/901#issuecomment-1153800207",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

File: eventmesh-registry-plugin/eventmesh-registry-consul/src/main/java/org/apache/eventmesh/registry/consul/service/ConsulRegistryService.java (hunk 1 above)
Comment (huyuanxin) [reply eventmesh-registry-plugin/eventmesh-registry-consul/src/main/java/org/apache/eventmesh/registry/consul/service/ConsulRegistryService.java:93]: Token support？

File: eventmesh-registry-plugin/eventmesh-registry-consul/src/main/java/org/apache/eventmesh/registry/consul/service/ConsulRegistryService.java (hunk 2 above)
Comment (huyuanxin) [reply eventmesh-registry-plugin/eventmesh-registry-consul/src/main/java/org/apache/eventmesh/registry/consul/service/ConsulRegistryService.java:151]: yep,I will delete it

File: eventmesh-registry-plugin/eventmesh-registry-consul/src/main/java/org/apache/eventmesh/registry/consul/service/ConsulRegistryService.java (hunk 1 above)
Comment (li-xiao-shuang) [reply eventmesh-registry-plugin/eventmesh-registry-consul/src/main/java/org/apache/eventmesh/registry/consul/service/ConsulRegistryService.java]: > Token support？

Yes, authentication needs to be turned on in actual applications, so it needs to support authentication

**Comment by xwm1992** [comment]
please solve the conflic files @huyuanxin .

**Comment by qqeasonchen** [comment]
@huyuanxin can you please fix the conflicts？

**Comment by huyuanxin** [comment]
> @huyuanxin can you please fix the conflicts？

fixing

**Comment by huyuanxin** [comment]
I cherry pick from latest upstream

**Review by li-xiao-shuang — APPROVED** [review]

**Comment by xwm1992** [comment]
I have created the `registry-consul` branch please merge this pr to this branch and we will fix the ci problems later. @huyuanxin

**Pull request merged**